File: app/src/main/java/me/kirimin/mitsumine/network/api/ApiAccessor.java
package me.kirimin.mitsumine.network.api;

import com.android.volley.RequestQueue;
import com.android.volley.toolbox.JsonObjectRequest;
import com.android.volley.toolbox.RequestFuture;
import com.android.volley.toolbox.StringRequest;

import org.json.JSONObject;

import java.util.concurrent.ExecutionException;

import me.kirimin.mitsumine.network.ApiRequestException;
import rx.Observable;
import rx.Subscriber;

class ApiAccessor {

    static Observable<JSONObject> apiRequest(final RequestQueue requestQueue, final String url) {
        return Observable.create(new Observable.OnSubscribe<JSONObject>() {
            @Override
            public void call(Subscriber<? super JSONObject> subscriber) {
                RequestFuture<JSONObject> future = RequestFuture.newFuture();
                requestQueue.add(new JsonObjectRequest(url, null, future, future));
                try {
                    JSONObject response = future.get();
                    subscriber.onNext(response);
                    subscriber.onCompleted();
                } catch (InterruptedException | ExecutionException e) {
                    subscriber.onError(new ApiRequestException(e.getMessage()));
                }
            }
        });
    }

    static Observable<String> stringRequest(final RequestQueue requestQueue, final String url) {
        return Observable.create(new Observable.OnSubscribe<String>() {
            @Override
            public void call(final Subscriber<? super String> subscriber) {
                RequestFuture<String> future = RequestFuture.newFuture();
                requestQueue.add(new StringRequest(url, future, future));
                try {
                    String response = future.get();
                    subscriber.onNext(response);
                    subscriber.onCompleted();
                } catch (InterruptedException | ExecutionException e) {
                    subscriber.onError(new ApiRequestException(e.getMessage()));
                }
            }
        });
    }
}

File: app/src/main/java/me/kirimin/mitsumine/network/api/ApiAccessor.kt
package me.kirimin.mitsumine.network.api

import com.squareup.okhttp.CacheControl
import com.squareup.okhttp.OkHttpClient
import com.squareup.okhttp.Request

import org.json.JSONObject
import org.scribe.builder.ServiceBuilder
import org.scribe.model.OAuthRequest
import org.scribe.model.Response
import org.scribe.model.Token

import me.kirimin.mitsumine.model.Account
import me.kirimin.mitsumine.network.ApiRequestException
import me.kirimin.mitsumine.network.api.oauth.Consumer
import me.kirimin.mitsumine.network.api.oauth.HatenaOAuthProvider
import rx.Observable
import java.util.concurrent.TimeUnit

class ApiAccessor {
    companion object {

        fun request(url: String): Observable<JSONObject> {
            return stringRequest(url).map { response ->
                JSONObject(response)
            }
        }

        fun stringRequest(url: String): Observable<String> {
            return Observable.create<String> { subscriber ->
                val client = OkHttpClient()
                val request = Request.Builder()
                        .url(url)
                        .cacheControl(CacheControl.Builder().maxAge(5, TimeUnit.MINUTES).build())
                        .get()
                        .build();
                val response = client.newCall(request).execute()
                if (response.isSuccessful) {
                    val json = response.body().string()
                    subscriber.onNext(json)
                    subscriber.onCompleted()
                } else {
                    subscriber.onError(ApiRequestException("error:" + response.code()));
                }
            }
        }

        fun oAuthRequest(account: Account, request: OAuthRequest): Response {
            val oAuthService = ServiceBuilder().provider(HatenaOAuthProvider::class.java).apiKey(Consumer.K).apiSecret(Consumer.S).build()
            val accessToken = Token(account.token, account.tokenSecret)
            oAuthService.signRequest(accessToken, request)
            return request.send()
        }
    }
}